Preview: Only pause DocumentTimeline animations in preview-web

tryggvigy opened the pull request:
Closes #31877

<!--

Thank you for contributing to Storybook! Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `main` branch as part of the release process, so you shouldn't need to worry about this. For additional guidance: https://storybook.js.org/docs/contribute

-->

## What I did

Change `pauseAnimations` in preview-web so that only DocumentTimeline animations are paused. `ScrollTimeline` and `ViewTimeline` animations are spared.

Unfortunately, it's impossible to target only a specific animation timeline type in pure CSS. This means there is a slight behaviour change:

**Nuke vs. Pause**
- Before, we set `animation: none !important` nuking the animation.
- After, we pause the animation at it's starting point.

**Subsequent animations**
- Before, the `pauseStyles` injected into the document, made sure to pause all animations that were added later in their initial state.
- After, the JS solution will only pause present animations. It won’t suppress future animations unless we track them and pause them dynamically.

I wonder if this would be considered a breaking change. I do still believe the current behaviour should change. As time goes on and more websites start to utilize View- and ScrollTimeline animations, this behaviour will become problematic for more users. 


## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

<!-- Please include the steps to test your changes here. For example:

1. Run a sandbox for template, e.g. `yarn task --task sandbox --start-from auto --template react-vite/default-ts`
2. Open Storybook in your browser
3. Access X story

-->

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

<!-- BENCHMARK_SECTION -->
<!-- BENCHMARK_SECTION -->


<!-- greptile_comment -->

## Greptile Summary

Improves animation handling in preview-web by selectively pausing only DocumentTimeline animations while preserving ScrollTimeline and ViewTimeline animations, switching from a CSS-based to Web Animations API approach.

- Switches from `animation: none !important` to using Web Animations API for more precise control
- Preserves ScrollTimeline and ViewTimeline animations that were previously incorrectly paused
- Changes animation state handling from 'nuking' to 'pausing' at starting point
- No longer automatically pauses future animations added after initial pause
- Potential breaking change for tests relying on previous animation behavior



<!-- /greptile_comment -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added animation demonstration stories to Storybook: fade-in, infinite rotation, scroll-triggered growth, and transition examples with interactive play scripts that simulate user interactions.

* **Improvements**
  * More reliable pause/restore of animations during accessibility checks so UI state is preserved.
  * Better animation handling across shadow DOM and per-animation control for accurate pausing and waiting.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tryggvigy committed:
Only pause DocumentTimeline animations in preview-web

ghengeveld commented:
Nice work @tryggvigy, you did almost exactly what I did, except that I hadn't committed anything yet 😅 

I'll update your branch with some aspects from mine (such as adding stories) and move forward with this PR so you'll get credited 👍

ghengeveld committed:
Refine pauseAnimations behavior and add stories to verify

ghengeveld committed:
Pause infinite animations on their first frame rather than canceling them

ghengeveld committed:
Pause and restore animations while running a11y tests. Update stories to demonstrate.

ghengeveld commented:
I iterated on this some more and found that it's actually worthwhile to restore animation state afterwards. The a11y addon can use this to pause infinite animations at their starting state in development mode, in order to prevent a discrepancy between a11y results in dev vs test. I also decided to pause infinite animations on their 1st frame rather than cancelling them outright, leading to a more predictable Chromatic snapshot and more accurate A11y results.

tryggvigy commented:
@ghengeveld I just got an idea that could possibly simplify things. We can access https://developer.mozilla.org/en-US/docs/Web/API/KeyframeEffect/target on the `Animation` instances returned from `document.getAnimations()` So we could keep the old CSS and the previous behaviour:

```ts
const animations = document
  .getAnimations()
 .filter(isDocumentAnimation)
 .forEach(animation => {
   animation.effect.target.setAttribute('data-disable-animation');
 });

...

  const pauseStyle = document.createElement('style');
  pauseStyle.textContent = `[data-disable-animation] {
    animation-delay: 0s !important;
    animation-direction: ${atEnd ? 'reverse' : 'normal'} !important;
    animation-play-state: paused !important;
    transition: none !important;
  }`;
  document.head.appendChild(pauseStyle);
```

I havent tested this yet. There might be some edge cases I'm missing, but it could help simplify the infinite animation headache.
Of course, just like our JS solution, this will also suffers from not being able to target future animations.

Which I hope is fine, you await your page being in the correct state before pausing animations. But it could be problematic if someone is writing e2e type tests via storybook and desire this behaviour (does storybook have that?). For example I could see a test that:

1. waits for a page to load
2. pauses animations
3. takes a screenshot
4. clicks a button that opens a modal
5. In the modal there is a newly mounted animation (maybe infinite animation)
6. Test can not take a stable screenshot.

Between step 5. and 6. animations will need to be paused again. But perhaps thats reasonable to expect test authors to do that ar writing these types of tests

ghengeveld commented:
Interesting concept, I'll give it some thought.
Animations getting added later is not a problem, that's what the event listeners are for (`animationstart` and `transitionrun`).

valentinpalkovic committed:
Merge branch 'next' into pause-document-timeline-animations-only

ghengeveld committed:
Merge branch 'next' into pause-document-timeline-animations-only

ghengeveld committed:
Move animationRoots inside pauseAllAnimations to ensure dynamically created shadow roots are handled

ghengeveld committed:
Use getTiming() instead of parsing computed animation names

ghengeveld committed:
Wrap animation restoration in try-catch to handle removed/invalid animations

ghengeveld committed:
fix(preview): only pause DocumentTimeline animations, leave scroll/view-driven untouched

Skip non-DocumentTimeline animations in pauseAnimations(): do not push them into previousStates or call pause/finish. Prevents scroll- and view-driven animations from being paused or rewound.

ghengeveld committed:
Call `animation.pause()` when the saved playState is 'paused', so each saved entry now restores both currentTime and its original play state.

valentinpalkovic committed:
Merge branch 'next' into pause-document-timeline-animations-only

ghengeveld committed:
Merge branch 'next' into pause-document-timeline-animations-only
